src/peft/tuners/lora/eva.py

```
@@ -0,0 +1,673 @@
+    values for later analysis or comparison.
+
+    Args:
+        name (str): Name of the layer to which this hook is attached. hashed_inputs (list): List of hashed inputs.
+        prepare_layer_inputs_fn (Optional[callable]): Function to prepare layer inputs for hashing.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        prepare_layer_inputs_fn: Optional[callable] = None,
+    ):
+        super().__init__(name, prepare_layer_inputs_fn)
+        self.hashed_inputs = []
+
+    @staticmethod
+    def hash_fn(tensor):
+        return hash(tuple(tensor.view(-1).tolist()))
+
+    @torch.no_grad()
+    def __call__(self, model, input, output):
+        x = self.prepare_layer_inputs(input).cpu()
+        self.hashed_inputs.append(self.hash_fn(x))
+
+
+def find_equal_values(dictionary: dict) -> dict:
+    """
+    Find keys in a dictionary that have the same value.
+
+    This function takes a dictionary and returns a new dictionary containing keys that have the same value. The keys in
+    the output dictionary are the values from the input dictionary, and the values are lists of keys that share the
+    same value.
+    """
+    value_dict = defaultdict(list)
+    for k, v in dictionary.items():
+        value_dict[v].append(k)
+    return {k: v for k, v in value_dict.items() if len(v) > 1}
+
+
+def get_target_name_key(name, pattern_keys):
+    """
+    Get the first substring of `name` that matches one of the keys in `pattern_keys`.
+    """
+    return next(filter(lambda key: re.match(rf".*\.{key}$", name), pattern_keys), name)
+
+
+def get_device_with_meta_params(model: torch.nn.Module) -> torch.device:
+    """
+    Get the device of the model's parameters. Useful if some parameters are on meta device.
+    """
+    devices = list({p.device for p in model.parameters() if p.device.type != "meta"})
+    if len(devices) > 1:
+        raise ValueError("model has multiple devices")
```

I wonder if we should really raise an error here. Maybe the user has some kind of custom model that can correctly deal with the devices, and raising here would only stop them unnecessarily. How about giving a warning and skipping the `move_inputs_to_device(inputs, device)` if `device` could not be determined?

---

sounds good